tiledb/sm/serialization/rest_version.cc

```
@@ -0,0 +1,194 @@
+ */
+
+#include <sstream>
+
+// clang-format off
+#ifdef TILEDB_SERIALIZATION
+#include <capnp/compat/json.h>
+#include <capnp/message.h>
+#include <capnp/serialize.h>
+#endif
+// clang-format on
+
+#include "tiledb/sm/enums/serialization_type.h"
+#include "tiledb/sm/rest/rest_client.h"
+#include "tiledb/sm/rest/rest_client_remote.h"
+#include "tiledb/sm/serialization/rest_version.h"
+#include "tiledb/sm/storage_manager/context.h"
+
+using namespace tiledb::common;
+
+namespace tiledb::sm::serialization {
+
+class RestVersionSerializationException : public StatusException {
+ public:
+  explicit RestVersionSerializationException(const std::string& message)
+      : StatusException("[TileDB::Serialization][RestVersion]", message) {
+  }
+};
+
+class RestVersionSerializationDisabledException
+    : public RestVersionSerializationException {
+ public:
+  explicit RestVersionSerializationDisabledException()
+      : RestVersionSerializationException(
+            "Cannot (de)serialize; serialization not enabled.") {
+  }
+};
+
+#ifdef TILEDB_SERIALIZATION
+
+void rest_version_to_capnp(
+    Context* ctx, capnp::RestVersion::Builder* rest_version_builder) {
+  if (ctx->has_rest_client()) {
+    // The REST version is initialized in the RestClientRemote constructor.
+    rest_version_builder->setTiledbVersion(
+        static_cast<RestClientRemote&>(ctx->rest_client()).rest_version());
+  } else {
+    // We should not hit this case, if TILEDB_SERIALIZATION is enabled there
+    // will be a rest client attached to ContextResources.
+    throw RestVersionSerializationException(
+        "Cannot serialize REST version with no initialized RESTClient.");
+  }
+}
+
+std::string rest_version_from_capnp(
+    const capnp::RestVersion::Reader& rest_version_reader) {
+  if (rest_version_reader.hasTiledbVersion()) {
+    return rest_version_reader.getTiledbVersion();
+  }
+  // TODO: What is best to return here?
+  return "2.X.0";
```

We could throw, that should be an error case. Alternatively we could default to the current tiledb version but I am afraid we could just be masking an error.

---

I was concerned if we change the field in a newer release older clients will not pass the `hasTiledbVersion()` check and always throw when making this request to REST since it would no longer provide the old field in it's response. :grimacing: I agree though and updated it to throw, it feels like a default value would mask the error and potentially even defeat the purpose of the endpoint (we can't assume REST's capabilities if this fails)